### DISCUSSION TITLE
Offical way to use Client Components

### DISCUSSION THREAD

**Discussion by mweel1** [remix-run/remix#8877] [Q&A]
What is the official way to use client components in remix.

I found this thread https://stackoverflow.com/questions/75706454/how-to-render-component-only-client-side-with-remix

I keep getting an error 
``` 
React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.
```
Using this code...
```import { LoaderFunctionArgs, json } from "@remix-run/node";
import { useLoaderData } from "@remix-run/react";
import { getContext, getRequest } from "@/utils";

import { ServiceOrder } from "@/models/serviceOrder";
import { lazy, Suspense } from "react";
import { ExpressCheckoutPay } from "@/components/expressCheckout/pay.client";
import { LoadingSpinner } from "ui/components/loadingSpinner";
import { Test } from "@/components/test.client";

export async function loader({ params, request }: LoaderFunctionArgs) {
  // get $id from route params
  const id = params.id;

  const r = await getRequest<ServiceOrder>(
    `/api/customer-portal/express-checkout/service-orders/${id}`,
    request
  );

  var context = await getContext(request);

  return json({
    serviceOrder: r,
    systemContext: context,
  });
}

export default function Page() {
  var loaderData = useLoaderData<typeof loader>();

  return (
    <>
      <Suspense>
        <ExpressCheckoutPay
          firstName={loaderData.serviceOrder.customer.firstName}
          lastName={loaderData.serviceOrder.customer.lastName}
          email={loaderData.serviceOrder.customer.email}
          serviceOrderId={loaderData.serviceOrder.serviceOrderId}
          systemContext={loaderData.systemContext}
          amountDue={loaderData.serviceOrder.amountDue}
        />
      </Suspense>
    </>
  );
}
```

if I remove ExpressCheckoutPay with some text it works fine.

**Comment by sergiodxa** [comment]
First, there's no RSC support in Remix, so every component is a client component.

I understand from your example that what you mean is a component that you export from a `.client` file, the `.client` extension means the file is not included in the server bundle, that's all, there's no difference with other components in Remix aside of that, contrarian to Next where a client component ("use client") will have access to hooks while a server component don't.

Now answering why you get that error, if you want to render a component a component that's only included in the client bundle and not in the server bundle, you need to delay the rendering to the browser, otherwise the component will be undefined during SSR and throw that error.

You can do this simply by using ClientOnly component from Remix Utils

```ts
import { ClientOnly } from "remix-utils/client-only"

<ClientOnly fallback={<SSRFallback />}>
  {() => {
    <YourComponent />
  })
</ClientOnly>
```

The SSRFallback is optional and default to null, which will cause content layout shift after hydration when the component inside is rendered, if that's not an issue you can skip the fallback.

If you don't want to install another library you can check the source code of how it works, the basic idea is that on SSR and hydration it renders the fallback, and then it change a state to trigger a new render and now it do render the component.

**Reply by kjkent** [reply]
@sergiodxa - I'm trying to understand, what's the point of having `.client` directories/files if it's still needed to use `if (typeof document !== 'undefined')`/`useEffect()`/`ClientOnly` to prevent an error?

**Reply by sergiodxa** [reply]
Some packages break on SSR just be importing them, e.g. let's say you have a module like this
```ts
let value = localStorage.getItem("key")
export function MyComponent() {
  return <p>{value}</p>
}
```
If you import this file in the server, it will crash, it will not even reach the render because just by importing it will crash, because a browser API (localStorage) is used at the module level.

I have found many times packages that do things like that, maybe to access `window` or `navigator` at the module level, instead of inside an effect.

That's where `.client` helps, you can re-export the component from a `.client` file, you will will need to delay the rendering to the browser but the module will not be included in the server.

**Reply by kjkent** [reply]
That's good info to know -- thank you for the explanation!

Edit: Also I just noticed you write `remix-utils`, which is helpful in wrapping my head around good remix practices and utility functions. Thanks again.

**Reply by jmjpro** [reply]
> First, there's no RSC support in Remix, so every component is a client component.
> 
> I understand from your example that what you mean is a component that you export from a `.client` file, the `.client` extension means the file is not included in the server bundle, that's all, there's no difference with other components in Remix aside of that, contrarian to Next where a client component ("use client") will have access to hooks while a server component don't.
> 
> Now answering why you get that error, if you want to render a component a component that's only included in the client bundle and not in the server bundle, you need to delay the rendering to the browser, otherwise the component will be undefined during SSR and throw that error.
> 
> You can do this simply by using ClientOnly component from Remix Utils
> 
> ```ts
> import { ClientOnly } from "remix-utils/client-only"
> 
> <ClientOnly fallback={<SSRFallback />}>
>   {() => {
>     <YourComponent />
>   })
> </ClientOnly>
> ```
> 
> The SSRFallback is optional and default to null, which will cause content layout shift after hydration when the component inside is rendered, if that's not an issue you can skip the fallback.
> 
> If you don't want to install another library you can check the source code of how it works, the basic idea is that on SSR and hydration it renders the fallback, and then it change a state to trigger a new render and now it do render the component.

I think you got your parentheses and curly braces confused there and it should be:
```ts
 <ClientOnly fallback={<SSRFallback />}>
   {() => (
     <YourComponent />
   )}
 </ClientOnly>
 ```

**Comment by GeMarVi** [comment]
If you're using a library in your Remix project that should only render on the client side, you can follow these steps:

Rename the filename where you import and use the library to end with .client.tsx.

For example, if you have a file named LibraryComponent.tsx, rename it to LibraryComponent.client.tsx.

Move the file to a .client directory. If you prefer to keep a cleaner directory structure, create a .client directory and move the LibraryComponent.tsx file inside that directory.

Update the imports in any file that references LibraryComponent to reflect the new filename or location of the file.
By making these changes, the LibraryComponent code will only be included in the client-side bundle, preventing it from running on the server. This approach is useful for libraries that rely on browser-specific objects or APIs.
